app/services/plugin.service.ts

```diff
@@ -2,139 +2,101 @@ import { Injectable, Inject } from '@angular/core';
+            if (!((st instanceof Type.ClassType) && (dt instanceof Type.ClassType))) {
+                throw "Validator.constructor: inconsistancy error";
+            }
+            this.edges.set(edgeType, [st, dt]);
+        }
     }
-    nodePluginData(type: string) {
-        return new PluginPropertyData(type, [
-            ["Start State", Type.Boolean],
-            ["Accept State", Type.Boolean],
-        ]);
+
+    isValidEdge(t: string, src: string, dst: string): boolean {
+        const edgets = this.edges.get(t);
+        const srct = this.nodes.get(src);
+        const dstt = this.nodes.get(dst);
+        if (!edgets || !srct || !dstt) {
+            throw "validator state error";
+        }
+        const [srce, dste] = edgets;
+        return srct.subtype(srce) && dstt.subtype(dste);
     }
-    edgePluginData(type: string) {
-        return new PluginPropertyData(type, []);
-    };
 }
 
+class ConcretePlugin implements Core.Plugin {
+    kind = "dfa.sinap.graph-kind";
 
-
-
-class MLPlugin implements Core.Plugin {
-    kind = "machine-learning.sinap.graph-kind";
-    nodeTypes = ["Input", "Fully Connected", "Conv2D", "Max Pooling", "Reshape", "Output"];
-    edgeTypes = ["Connection"];
-
-    validator = {
-        isValidEdge(t: string, src: string, dst: string) {
-            return true;
-        }
+    get nodeTypes() {
+        return this.definitions.nodes.keys();
     }
-
-    graphPluginData() {
-        return new PluginPropertyData("Graph", []);
+    get edgeTypes() {
+        return this.definitions.nodes.keys();
     }
 
-    nodePluginData(type: string) {
-        return new PluginPropertyData(type, this.nodePluginDataHelper(type));
+    validator: Validator;
+
+    constructor(private definitions: Definitions, public script: Script) {
+        this.validator = new Validator(definitions);
     }
 
-    private nodePluginDataHelper(type: string): [string, Type.Type][] {
-        switch (type) {
-            case "Input":
-                return [["shape", Type.String]];
-            case "Fully Connected":
-                return [];
-            case "Conv2D":
-                return [["stride", Type.Point],
-                ["output depth", Type.Number]];
-            case "Max Pooling":
-                return [["size", Type.Point]];
-            case "Reshape":
-                return [["shape", Type.String]];
-            case "Output":
-                return [];
-            default:
-                return [["beta", Type.Boolean]];
+    graphPluginData(kind: string) {
+        const type = this.definitions.graphs.get(kind);
+        if (!type) {
+            throw "type not found";
         }
+        return new PluginPropertyData(kind, type);
     }
-
-    edgePluginData(type: string) {
-        return new PluginPropertyData("Edge", []);
+    nodePluginData(kind: string) {
+        const type = this.definitions.nodes.get(kind);
```

Comment:
nodePluginData et. al. should have a hidden private method that accepts kind and a Map since they are basically the same and seem to differ only in which dictionary is used.